promql/prettier/prettier.go

```diff
@@ -0,0 +1,480 @@
+package prettier
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	"io/ioutil"
+	"reflect"
+	"regexp"
+)
+
+const (
+	// PrettifyRules for prettifying rules files along with the
+	// expressions in them.
+	PrettifyRules = iota
+	// PrettifyExpression for prettifying instantaneous expressions.
+	PrettifyExpression
+)
+
+// Prettier handles the prettifying and formatting operation over a
+// list of rules files or a single expression.
+type Prettier struct {
+	files      []string
+	expression string
+	Type       uint
+	Node       parser.Expr
+	pd         *padder
+}
+
+type padder struct {
```

Comment:
Do you really need to carry around this much state?

Reply:
Try to reduce the amount of state as much as possible and rather look up things in the syntax tree on demand.

Reply:
I think, if we don't save that information, then the look-ups in AST would be much higher. This will result in an exponential complexity. However, this does not mean that the current approach is efficient. More work will be put into this in terms of efficiency.

Reply:
> This will result in an exponential complexity.

Do you mean literally exponential complexity?

I don't see how it could get worse than O(n^2), which would be fine since that is already the complexity of the parser.

In general you should avoid such premature optimisation.

Reply:
> I don't see how it could get worse than O(n^2)

How about an expression consisting of only recursive binary-expressions with vector-selectors?

> In general you should avoid such premature optimisation.

Sure.

Reply:
> How about an expression consisting of only recursive binary-expressions with vector-selectors?

That should still be in `O(n^2)`.

Reply:
If you're interested why, you can imagine the following pseudo algorithm that should be slower than what we are doing. 

Let `n` be the number of bytes in the query. Then there are at most `n` Items and at most `n` AST nodes.

For each of the at most `n` items:
  * Print at least `n` bytes (when printing the Item).
  * Visit each of the `n` nodes at most 10 times.
  * Perform some tests regarding line length which include generating a string with length up to `n` and visiting at least `n` nodes.

This leaves you with a complexity of `O(n*(n+10*n+n))` which can be simplified to `O(n^2)`.

Reply:
Got it. Thank you for the explanation 😀.

Reply:
These are some information to keep the prettier aware of the surrounding items and their state. I think they are necessary, since they make the life easy for prettier.

Reply:
@slrbtfs, removed all those vars from `padder` that are not dependent on previous padder states/items. Now, they are being fetched from the node information instead of carrying in the padder. Only those that are dependent on previous items or are dependent across multiple items are being carried in padder.

Reply:
If this seems right, please resolve the comment.

Reply:
> If this seems right, please resolve the comment.

I can't, since I don't have the access rights for that.

But, the changes LGTM.

Reply:
> I can't, since I don't have the access rights for that.

Oh, I didn't know that.


> But, the changes LGTM.

Sure. Resolving.